packages/@aws-cdk/aws-codebuild/lib/build-spec.ts

```diff
@@ -0,0 +1,101 @@
+import { IResolveContext, Lazy, Stack } from '@aws-cdk/cdk';
+
+/**
+ * BuildSpec for CodeBuild projects
+ */
+export abstract class BuildSpec {
+  public static fromObject(value: {[key: string]: any}): BuildSpec {
+    return new StructuredBuildSpec(value);
+  }
+
+  /**
+   * Use a file from the source as buildspec
+   *
+   * Use this if you want to use a file different from 'buildspec.yml'`
+   */
+  public static fromFilename(filename: string): BuildSpec {
+    return new FilenameBuildSpec(filename);
+  }
+
+  protected constructor() {
+  }
+
+  /**
+   * Render the represented BuildSpec
+   */
+  public abstract toBuildSpec(): string;
+}
+
+/**
+ * BuildSpec that just returns the input unchanged
+ */
+class FilenameBuildSpec extends BuildSpec {
+  constructor(private readonly filename: string) {
+    super();
+  }
+
+  public toBuildSpec(): string {
+    return this.filename;
+  }
+
+  public toString() {
+    return `<buildspec file: ${this.filename}>`;
+  }
+}
+
+/**
+ * BuildSpec that understands about structure
+ */
+class StructuredBuildSpec extends BuildSpec {
+  constructor(public readonly spec: {[key: string]: any}) {
+    super();
+  }
+
+  public toBuildSpec(): string {
+    // We have to pretty-print the buildspec, otherwise
+    // CodeBuild will not recognize it as an inline buildspec.
+    return Lazy.stringValue({ produce: (ctx: IResolveContext) =>
+      Stack.of(ctx.scope).toJsonString(this.spec, 2)
```

Comment:
why does this need to be lazy?

Reply:
Because we need access to `stack.toJsonString`. 

`CloudFormationLang.toJson()` would have been good enough, but that's internal and hidden so I can't get to it.